### ISSUE TITLE
Support server side value conversions

### ISSUE THREAD

**Issue by ajcvickers** [aspnet/EntityFrameworkCore#10861]
Value conversions as implemented for #242 work by converting the value after it has come from the database and before it is send back to the database. In some situations it is useful to instead convert in the generated SQL. This enables support for cases where the database provider has no support for some database type, but it can be converted to some other database type that the provider can handle.

**Comment by ajcvickers** [comment]
Also consider the discussion in #12762 when working on this.

**Comment by roji** [comment]
Am interested in the reason for punting this - have you guys found a different way of implementing spatial support for SQL Server?

This would be useful for some other uses - PostgreSQL has some types which the Npgsql ADO driver doesn't support. These are mainly exotic/rare types, but server-side conversions would allow converting them to text before fetching. Definitely nothing important that should change the priority of this, though.

**Comment by bricelam** [comment]
> Have you guys found a different way of implementing spatial support for SQL Server?

Yes. We're reading and writing bytes directly based on [the serialization format](https://msdn.microsoft.com/en-us/library/ee301648(v=sql.105).aspx).

**Comment by roji** [comment]
Oh, OK... Does this mean the .NET Core SqlClient will soon be able to read and write SqlGeometry/Geography? Or is this something you're implementing at the EF Core level?

**Comment by ajcvickers** [comment]
@roji EF level, unfortunately.

**Comment by bricelam** [comment]
Well... we've implemented an NTS I/O reader and writer. Anyone could use them directly with SqlDataReader and SqlParameter. But obviously this is a lot different than using SqlGeometry directly on the client.

**Comment by roji** [comment]
@dpsenner translating operations on field of PostgreSQL composite types (UDTs) is a different thing from what this issue is about, and AFAIK requires Npgsql support rather than anything specific from EF Core.

This issue is about translating done database type to something else in SQL, before it is fetched to the client (and accordingly for sending).

**Comment by dpsenner** [comment]
Sorry, comment removed.

**Comment by ajcvickers** [comment]
@smitpatel See comment above, which is essentially about SQL translation with value converters. This is likely a duplicate of something, but I can't find the best bit.

**Comment by smitpatel** [comment]
@ajcvickers @tystol - Duplicate of #12045

**Comment by marchy** [comment]
Given [#12045](https://github.com/dotnet/efcore/issues/12045) is now closed, does this mean this is as well?

**Comment by ajcvickers** [comment]
@marchy Nope, this is still in the backlog for consideration.

**Comment by JohnYoungers** [comment]
In its current state, is there a hook/interceptor strategy that could be utilized to manually resolve this on a per usage basis?

If I have a postgres column that needs `encode(pgp_sym_encrypt('value', 'my_passphrase'), 'base64')` when stored and `pgp_sym_decrypt(decode(my_column, 'base64'), 'my_passphrase')` when queried, is there a way to slip that into the query?

**Comment by roji** [comment]
@JohnYoungers not automatically.

For query specifically, you can have .NET methods stubs that you call inside your LINQ queries (for encode, pgp_sym_encrypt...), and then map these to the appropriate PG function. It's not as nice since you have to manually include these calls in every query, but it should work.

But for SaveChanges I don't think there's a way to make this work, except maybe hacking it via a command interceptor and doing search-and-replace on the actual INSERT/UPDATE SQL (which is brittle and generally discouraged - but might be a workaround here).
